server/src/index.ts

```diff
@@ -1,29 +1,103 @@
+const Auth0JKWS = jose.createRemoteJWKSet(
+  new URL(`${auth0Tenant}.well-known/jwks.json`)
+);
+
 serve(
-  async (request) => {
-    if (request.method === "POST") {
-      const { method, params } = await request.json();
-      if (method === "get") {
-        const { key, id } = params;
-        return new Response((await redisGet(id + key)) || "null");
-      }
-      if (method === "set") {
-        const { key, value, id, ttl } = params;
-        const ttlOrDefault = ttl || oneWeekInSeconds;
-        await redisSet(
-          id + key,
-          value,
-          ttlOrDefault > oneWeekInSeconds ? oneWeekInSeconds : ttlOrDefault,
-        );
-        return new Response(JSON.stringify({}));
-      }
-    }
-    return new Response();
-  },
-  { port: parseInt(Deno.env.get("PORT") as string) },
+  app({
+    "/": {
+      POST: authenticated(api_token)(async (request, auth) => {
+        const { method, params } = await request.json();
+        if (method === "get") {
+          const { key } = params;
+          return new Response((await redisGet(`${auth.uid}:${key}`)) || "null");
+        }
+        if (method === "set") {
+          const { key, value, ttl } = params;
+          const ttlOrDefault = ttl || oneWeekInSeconds;
+          await redisSet(
+            `${auth.uid}:${key}`,
+            value,
+            ttlOrDefault > oneWeekInSeconds ? oneWeekInSeconds : ttlOrDefault
+          );
+          return new Response(JSON.stringify({}));
+        }
+        return new Response("Unknown method", { status: 400 });
+      }),
+    },
+    "/api-token/": {
+      GET: authenticated(auth0)(async (_request, auth) => {
+        const token = await redisClient.get(`user-api-token:${auth.uid}`);
+        if (token == null) {
+          return Response404();
+        }
+        return new Response(token);
+      }),
+      POST: authenticated(auth0)(async (_request, auth) => {
+        const old_token = await redisClient.get(`user-api-token:${auth.uid}`);
+        if (old_token != null) {
+          await redisClient.del(`api-token:${old_token}`);
+        }
+
+        const token = crypto.randomUUID();
+        await redisClient.set(`user-api-token:${auth.uid}`, token);
+        await redisClient.set(`api-token:${token}`, auth.uid);
+        return new Response(token);
+      }),
+    },
+  }),
+  { port: parseInt(Deno.env.get("PORT") as string) }
 );
+
+function getBearer(request: Request) {
+  return request.headers.get("Authorization")?.split("Bearer ")[1];
+}
+
+export async function auth0(request: Request) {
+  const jwt = getBearer(request);
+  if (!jwt) {
+    return null;
+  }
+
+  const { payload, protectedHeader } = await jose.jwtVerify(jwt, Auth0JKWS, {
+    issuer: auth0Tenant,
+    audience: "rmmbr",
+  });
+
+  const uid = payload.sub;
+  if (!uid) {
+    return null;
+  }
+
+  return {
+    jwt,
+    uid,
+    payload,
+    protectedHeader,
+  };
+}
+
+export async function api_token(request: Request) {
```

Comment:
consider improving on the function name, i.e. `requestToUserId` or something like that